### ISSUE TITLE
cargo test / bench much slower than rustc --test

### ISSUE THREAD

**Issue by mahkoh** [rust-lang/cargo#1394]
| command | compile time | 
|----------------|------------------|
| rustc --test -O | 60s |
| cargo bench --no-run | 66s |
| rustc --test | 22s |
| cargo test --no-run | 33s |

The reason that it's not much worse is that cargo builds things in parallel. Unfortunately I can't demonstrate `rustc --test -C codegen-units=2` because it gives me an assert.

**Comment by alexcrichton** [comment]
I believe this should probably be coupled with a little more analysis (or a reproduction would also be helpful). Cargo passes `-g` for example with `cargo test` which you do not indicate to have passed. Additionally cargo runs `-C opt-level=3` which you also do not indicate you have passed. I would recommend passing `-v` to cargo to learn the exact command line being passed and start comparing there.

**Comment by mahkoh** [comment]
@alexcrichton It's because cargo compiles the library and the test binary when you call cargo test. Since they are compiled in parallel the compile time doesn't explode but it prevents things such as using multiple codegen units.

**Comment by mahkoh** [comment]
`-g` increases the compile time for --test to `26 seconds`.

**Comment by mahkoh** [comment]
Compiling the library is necessary for doc tests but that's not what you want 99% of the time.

**Comment by alexcrichton** [comment]
Ah in that case I don't think there's actually a bug here, if you don't want the tests compiled for a library or a binary you can have `test = false` in the manifest.

**Comment by mahkoh** [comment]
Who said anything about not wanting to run tests when running cargo test?

**Comment by huonw** [comment]
Sounds like you just want a way to opt-out of doc tests?

**Comment by mahkoh** [comment]
Anything that requires the library. E.g. the things in test/ probably also needs it.

**Comment by huonw** [comment]
Feel free to open another issue that is actually clear about what is desired.

**Comment by mahkoh** [comment]
Thanks, but I've already chosen to simply add the desired option to my fork.
